slack_ros/scripts/slack_ros.py

```diff
@@ -34,46 +34,72 @@ def __init__(self):
         self.username = rospy.get_param('~username', 'ros-bot')
 	
         # Create a publisher for our custom message.
-        pub = rospy.Publisher('from_slack_to_ros', String, queue_size=10)
-	rospy.Subscriber("from_ros_to_slack", String, self.callback)
-	rospy.Subscriber("send_file_to_slack", String, self.filecallback)
+        self.pub = rospy.Publisher('from_slack_to_ros', String, queue_size=10)
+        rospy.Subscriber("from_ros_to_slack", String, self.message_callback)
+        rospy.Subscriber("send_file_to_slack", String, self.file_callback)
 
-	# Create the slack client
-	self.sc = SlackClient(self.token)
+        # Create the slack client
+        self.slack_client = SlackClient(self.token)
+        self.has_connection = False
+        self.subscribed_msg_store = []
```

Comment:
You could use a condition variable for this. Just wait/block inside the topic callback until the condition is fulfilled (connected to slack). https://docs.python.org/2/library/threading.html
With this kind of implementation you need some kind of connectivity check routine.

On the other side. Creating a separate queue like you did isn't that wrong. As long as there are elements within the queue you can try sending them. If an error during sending happens you can try reconnect and leave the element inside the queue as long as the connection problems appears...

@ipa-fmw what do you think?